### PR TITLE
ESQL: add =~ operator (case insensitive equality)

### PR DESCRIPTION
**Pull request by luigidellaquila**
Add an operator to perform case insensitive equality comparison of strings.
The operator symbol is `=~`, eg.

```
row a = "FooBar" | where a =~ "foobar"
```

~~The operator acts as a normal `==` for all non-string types~~
The operator is defined only for string values

For multi-values the behavior is consistent with `==`, ie. `multi_value =~ "string"` returns `null`

Related to https://github.com/elastic/elasticsearch/issues/103599

### CODE REVIEW

File: x-pack/plugin/esql/qa/testFixtures/src/main/resources/ignore-case.csv-spec [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+
+emp_no:integer | first_name:keyword | last_name:keyword
+10001          | Georgi             | Facello
+10011          | Mary               | Sluis
+;
+
+
+evalEquals#[skip:-8.12.99, reason:case insensitive operators implemented in v 8.13]
+from employees | where emp_no == 10001 
+| eval a = first_name =~ "georgi", b = first_name == "georgi", c = first_name =~ "GEORGI", d = first_name =~ "Geor"
```
Comment (costin) [inline x-pack/plugin/esql/qa/testFixtures/src/main/resources/ignore-case.csv-spec:42]: try "GeoRgI" or something similar just in cae.

File: x-pack/plugin/esql/src/main/antlr/EsqlBaseLexer.g4 [hunk 2]
```diff
@@ -130,6 +130,7 @@ RP : ')';
 TRUE : 'true';
 
 EQ  : '==';
+EQ_IGNORE_CASE  : '=~';
```
Comment (costin) [inline x-pack/plugin/esql/src/main/antlr/EsqlBaseLexer.g4:133]: I propose we use SEQ for "string equality" (just like in EQL)  or IEQ for "Insensitive Equality".
Same for the backing class.

File: x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/expression/predicate/operator/comparison/EqualsIgnoreCase.java [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+import org.elasticsearch.xpack.ql.expression.Expression;
+import org.elasticsearch.xpack.ql.expression.TypeResolutions;
+import org.elasticsearch.xpack.ql.tree.NodeInfo;
+import org.elasticsearch.xpack.ql.tree.Source;
+
+import java.time.ZoneId;
+
+import static org.elasticsearch.xpack.ql.expression.TypeResolutions.ParamOrdinal.DEFAULT;
+
+public class EqualsIgnoreCase extends BinaryComparison {
```
Comment (costin) [inline x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/expression/predicate/operator/comparison/EqualsIgnoreCase.java:18]: StringEquals or InsensitiveEquals

File: x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/expression/predicate/operator/comparison/Comparisons.java [hunk 4]
```diff
@@ -25,6 +25,10 @@ public static Boolean eq(Object l, Object r) {
         return i == null ? null : i.intValue() == 0;
     }
 
+    public static Boolean eqIgnoreCase(Object l, Object r) {
```
Comment (costin) [inline x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/expression/predicate/operator/comparison/Comparisons.java:28]: Who's using this method?
If needed see StringComparisons inside EQL

File: x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/expression/predicate/operator/comparison/EqualsIgnoreCase.java [hunk 5]
```diff
@@ -0,0 +1,48 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+package org.elasticsearch.xpack.ql.expression.predicate.operator.comparison;
```
Comment (costin) [inline x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/expression/predicate/operator/comparison/EqualsIgnoreCase.java:7]: This should be in the ESQL package.

File: x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/planner/ExpressionTranslators.java [hunk 6]
```diff
@@ -336,15 +337,15 @@ static Query translate(BinaryComparison bc, TranslatorHandler handler) {
             if (bc instanceof LessThanOrEqual) {
                 return new RangeQuery(source, name, null, false, value, true, format, zoneId);
             }
-            if (bc instanceof Equals || bc instanceof NullEquals || bc instanceof NotEquals) {
+            if (bc instanceof Equals || bc instanceof NullEquals || bc instanceof NotEquals || bc instanceof EqualsIgnoreCase) {
```
Comment (costin) [inline x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/planner/ExpressionTranslators.java:340]: No need to touch QL - see InsensitiveBinaryComparisons inside EQL on how to extend the method (add a dedicated TranslatorHandler).

**Review by costin — CHANGES_REQUESTED** [review]
No need to touch the QL side - see my comments.
Also the operator need to work on multi-value fields - please add unwrapping and tests.

File: x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/expression/predicate/operator/comparison/Comparisons.java (hunk 4 above)
Comment (luigidellaquila) [reply x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/expression/predicate/operator/comparison/Comparisons.java:28]: Removed (together with all the other changes in QL)

**Comment by luigidellaquila** [comment]
Thanks for the feedback @costin, I pushed some changes move all the logic in ESQL module (apart from one small fix in QL).

> Also the operator need to work on multi-value fields

I think this aspect needs some clarifications, let's move the discussion off-line

**Review by astefan — COMMENTED** [review]
What's the expected behavior when doing something like `first_name =~ last_name`?
The original issue this PR covers says
> Introduce =~ operator for performing case insensitive equality between a string field and a string literal
The field can be both single and multi value. The literal has to be single value.

Should I see an error, or warning or did the requirements change in the meantime and the original issue and/or the PR description needs an update?

---

On a similar note, if I do `where last_name =~ 123` I get no results and no relevant warning. What is the expected behavior in this case?

---

Another similar behavior is with `where last_name =~ languages` where I get a CCE:

```
class org.elasticsearch.compute.data.IntArrayBlock cannot be cast to class org.elasticsearch.compute.data.BytesRefBlock
```

but if I do `where last_name == languages` I am getting a more meaningful error message:

```
"reason": "Found 1 problem\nline 1:24: first argument of [last_name == languages] is [keyword] so second argument must also be [keyword] but was [integer]",
                "stack_trace": "org.elasticsearch.xpack.esql.analysis.VerificationException: Found 1 problem\nline 1:24: first argument of [last_name == languages] is [keyword] so second argument must also be [keyword] but was [integer]
```

**Comment by astefan** [comment]
This test also fails `row last_name = "lortz" | where "lOrTz" =~ last_name`:

```
org.elasticsearch.xpack.ql.QlIllegalArgumentException: Should not fold expression\r\n\tat org.elasticsearch.xpack.ql.expression.Expression.fold(Expression.java:82)\r\n\tat org.elasticsearch.xpack.ql.expression.Literal.of(Literal.java:110)\r\n\tat org.elasticsearch.xpack.ql.optimizer.OptimizerRules$ConstantFolding.rule(OptimizerRules.java:101)
```

whereas `row last_name = "lortz" | where "lortz" == last_name` doesn't.

**Comment by luigidellaquila** [comment]
> What's the expected behavior when doing something like first_name =~ last_name?

I don't have a strong opinion on this.
I supported it in the implementation just because also == supports it, but there is an optimization in case the right-hand side is foldable.
On the other hand, `LIKE` and `RLIKE` accept only a literal, so =~ could be the same.

For sure having a string literal there will reduce validation risks, solving possible problems at parsing time.

> On a similar note, if I do where last_name =~ 123 I get no results and no relevant warning

I think it's a bug (lack of validation), it should throw an exception as == does.

> Another similar behavior is with where last_name =~ languages where I get a CCE:

same as above. I'll check and fix it.

> This test also fails row last_name = "lortz" | where "lOrTz" =~ last_name:

this is interesting, I'll investigate

Thanks!

**Comment by astefan** [comment]
> > What's the expected behavior when doing something like first_name =~ last_name?
> 
> I don't have a strong opinion on this. I supported it in the implementation just because also == supports it, but there is an optimization in case the right-hand side is foldable. On the other hand, `LIKE` and `RLIKE` accept only a literal, so =~ could be the same.

If this is supported on purpose, then there should be tests covering it.
But, at the same time, the original issue mentions a literal. My questions spawned from this confusion between the issue description and the implementation itself which should be clarified.

**Comment by luigidellaquila** [comment]
Thanks for you feedback @astefan 
I pushed two fixes and more tests to cover validation and field-to-field comparison.

**Comment by luigidellaquila** [comment]
@elasticmachine update branch

**Comment by nik9000** [comment]
I hear tell this is waiting on me. I'll review as soon as I can.

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/planner/EsqlTranslatorHandler.java [hunk 7]
```diff
@@ -56,4 +104,83 @@ public Query wrapFunctionQuery(ScalarFunction sf, Expression field, Supplier<Que
...
+            }
+
+            String name = pushableAttributeName(attribute);
+
+            Query query;
+            if (isDateLiteralComparison) {
+                // dates equality uses a range query because it's the one that has a "format" parameter
+                query = new RangeQuery(source, name, value, true, value, true, format, zoneId);
+            } else {
+                query = new TermQuery(source, name, value, DataTypes.isString(attribute.dataType()));
```
Comment (nik9000) [inline x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/planner/EsqlTranslatorHandler.java:180]: I think `version` fields probably do want to push down case insensitive comparison.

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/evaluator/predicate/operator/comparison/InsensitiveEquals.java [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+    }
+
+    @Evaluator(extraName = "Doubles")
+    static boolean processDoubles(double lhs, double rhs) {
+        return lhs == rhs;
+    }
+
+    @Evaluator(extraName = "Keywords")
+    static boolean processKeywords(BytesRef lhs, BytesRef rhs) {
+        return lhs.utf8ToString().equalsIgnoreCase(rhs.utf8ToString());
```
Comment (nik9000) [inline x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/evaluator/predicate/operator/comparison/InsensitiveEquals.java:58]: I don't believe this'll work for `BytesRef`. I think that supports non-utf8 text.

I'm also not 100% confident `equalsIgnoreCase` will have the same behavior as the Lucene queries that come out of the field mappers, though I haven't checked. if you have then cool.

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/evaluator/predicate/operator/comparison/InsensitiveEquals.java [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+    protected NodeInfo<InsensitiveEquals> info() {
+        return NodeInfo.create(this, InsensitiveEquals::new, left(), right(), zoneId());
+    }
+
+    @Override
+    protected InsensitiveEquals replaceChildren(Expression newLeft, Expression newRight) {
+        return new InsensitiveEquals(source(), newLeft, newRight, zoneId());
+    }
+
+    @Evaluator(extraName = "Ints")
```
[the remark this answers is not in the document]
Comment (luigidellaquila) [reply x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/evaluator/predicate/operator/comparison/InsensitiveEquals.java:41]: Yeah, I don't have a strong opinion on this. I just implemented it as it was described in the initial issue https://github.com/elastic/elasticsearch/issues/103599.

And yes, I think I can simplify it an in case use Equals implementation

File: x-pack/plugin/esql/src/main/antlr/EsqlBaseLexer.g4 [hunk 10]
```diff
@@ -130,6 +130,7 @@ RP : ')';
 TRUE : 'true';
 
 EQ  : '==';
+SEQ  : '=~';
```
[the remark this answers is not in the document]
Comment (luigidellaquila) [reply x-pack/plugin/esql/src/main/antlr/EsqlBaseLexer.g4:133]: My first proposal was for `EQ_IGNORE_CASE`, but it was a bit verbose.
We went for `SEQ` only because it's the same we have in EQL

File: x-pack/plugin/esql/qa/testFixtures/src/main/resources/ignore-case.csv-spec [hunk 11]
```diff
@@ -0,0 +1,63 @@
+
+simpleFilter#[skip:-8.12.99, reason:case insensitive operators implemented in v 8.13]
```
[the remark this answers is not in the document]
Comment (luigidellaquila) [reply x-pack/plugin/esql/qa/testFixtures/src/main/resources/ignore-case.csv-spec:2]: It hurts me as well, especially when I see it in the logs. But does it work if I move it to the next line? I'll check it separately

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/evaluator/predicate/operator/comparison/InsensitiveEquals.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (nik9000) [reply x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/evaluator/predicate/operator/comparison/InsensitiveEquals.java:41]: @costin do you think `=~` should allow comparisons with non-string types? I think maybe no?

File: x-pack/plugin/esql/qa/testFixtures/src/main/resources/ignore-case.csv-spec (hunk 11 above)
[the remark this answers is not in the document]
Comment (nik9000) [reply x-pack/plugin/esql/qa/testFixtures/src/main/resources/ignore-case.csv-spec:2]: Yeah. Separately.

File: x-pack/plugin/esql/src/main/antlr/EsqlBaseLexer.g4 (hunk 10 above)
[the remark this answers is not in the document]
Comment (costin) [reply x-pack/plugin/esql/src/main/antlr/EsqlBaseLexer.g4:133]: I can see how SEQ can be confusing. CEIQ or EQS (Equals String) both work for me.
I prefer string equality since it's more generic and it might go beyond just case insensitive comparisons.

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/evaluator/predicate/operator/comparison/InsensitiveEquals.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (costin) [reply x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/evaluator/predicate/operator/comparison/InsensitiveEquals.java:41]: +1 on falling back to the regular Equals operators for non string params.

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/planner/EsqlTranslatorHandler.java [hunk 12]
```diff
@@ -56,4 +104,83 @@ public Query wrapFunctionQuery(ScalarFunction sf, Expression field, Supplier<Que
...
+
+            ZoneId zoneId = null;
+            if (DataTypes.isDateTime(attribute.dataType())) {
+                zoneId = bc.zoneId();
+            }
+
+            String name = pushableAttributeName(attribute);
+
+            Query query;
+            if (isDateLiteralComparison) {
```
[the remark this answers is not in the document]
Comment (costin) [reply x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/planner/EsqlTranslatorHandler.java:176]: The operator should fallback to equals for other type so the other types should not make a difference.
IPs fields would be handled separately even when dealing with a string value:

`ip_field =~ "2001:0:5ef5..."` should be the same as `ip_field == "2001:0:5ef5..."`

The  IPv6 comparisons itself should case-insensitive but at this point we're discussing about IP fields not =~

[the remark this answers is not in the document]
Comment (nik9000) [reply x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/planner/EsqlTranslatorHandler.java:176]: Yeah - ipv6 is case insensitive all the time because we're just parsing to the ip. It's insensitive of normalization too.

I dunno, I kind of feel like `=~` just shouldn't work on non-string fields.

[the remark this answers is not in the document]
Comment (costin) [reply x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/planner/EsqlTranslatorHandler.java:176]: >  I kind of feel like =~ just shouldn't work on non-string fields.
Makes it easier to reason about so let's do that and throw a validation error in other cases.

File: x-pack/plugin/esql/src/main/antlr/EsqlBaseLexer.g4 (hunk 10 above)
[the remark this answers is not in the document]
Comment (luigidellaquila) [reply x-pack/plugin/esql/src/main/antlr/EsqlBaseLexer.g4:133]: Ok, I see a majority for CIEQ, I'll use it

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/planner/EsqlTranslatorHandler.java (hunk 7 above)
Comment (luigidellaquila) [reply x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/planner/EsqlTranslatorHandler.java:180]: No version support for now (see above)

**Comment by luigidellaquila** [comment]
@elasticmachine run elasticsearch-ci/part-1

**Comment by luigidellaquila** [comment]
Folks, I think this is in a good state for another round of reviews. 
It also includes wildcard support now.

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/planner/EsqlTranslatorHandler.java [hunk 13]
```diff
@@ -56,4 +92,38 @@ public Query wrapFunctionQuery(ScalarFunction sf, Expression field, Supplier<Que
...
+                bc.symbol()
+            );
+        }
+
+        static Query translate(InsensitiveEquals bc) {
+            TypedAttribute attribute = checkIsPushableAttribute(bc.left());
+            Source source = bc.source();
+            BytesRef value = BytesRefs.toBytesRef(ExpressionTranslators.valueOf(bc.right()));
+            String name = pushableAttributeName(attribute);
+            Query query = new WildcardQuery(source, name, value.utf8ToString(), true);
```
Comment (costin) [inline x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/planner/EsqlTranslatorHandler.java:124]: This is incorrect - the equality should translate to a Term query with case_insensitive. The wildcard query is used only by LIKE.

**Review by costin — CHANGES_REQUESTED** [review]
This PR needs more work.

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/planner/EsqlTranslatorHandler.java (hunk 13 above)
Comment (luigidellaquila) [reply x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/planner/EsqlTranslatorHandler.java:124]: Maybe I got it completely wrong, but from the original [issue](https://github.com/elastic/elasticsearch/issues/103599) and from the off-line discussions my understanding was that we wanted wildcard expressions on the right-hand side. 
If we want it, then we'll need `WildcardQuery` here.

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/evaluator/predicate/operator/comparison/InsensitiveBinaryComparison.java [hunk 14]
```diff
@@ -0,0 +1,34 @@
...
+import java.time.ZoneId;
+
+public abstract class InsensitiveBinaryComparison extends BinaryScalarFunction {
+
+    private final ZoneId zoneId;
+
+    protected InsensitiveBinaryComparison(Source source, Expression left, Expression right, ZoneId zoneId) {
+        super(source, left, right);
+        this.zoneId = zoneId;
+    }
```
[the remark this answers is not in the document]
Comment (luigidellaquila) [reply x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/evaluator/predicate/operator/comparison/InsensitiveBinaryComparison.java:24]: Yeah, that's a leftover from the first implementation that also supported date types.
I'm dropping it.

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Verifier.java [hunk 15]
```diff
@@ -260,6 +268,15 @@ private void gatherMetrics(LogicalPlan plan, BitSet b) {
...
     public static Failure validateBinaryComparison(BinaryComparison bc) {
+        return validateBinaryComparison((BinaryScalarFunction) bc);
+    }
+
+    public static Failure validateBinaryComparison(InsensitiveBinaryComparison bc) {
+        return validateBinaryComparison((BinaryScalarFunction) bc);
+    }
+
+    private static Failure validateBinaryComparison(BinaryScalarFunction bc) {
+        assert bc instanceof BinaryComparison || bc instanceof InsensitiveBinaryComparison;
```
[the remark this answers is not in the document]
Comment (luigidellaquila) [reply x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Verifier.java:279]: Do we want do to it now?
I implemented this validation consistently with other `BinaryComparison` operators, but I agree that it would be better to have the validation (and the resolution, that is a different aspect) inside the single operator classes.
IMHO it would be better to be consistent now, and then do a consistent refactoring for all the operators

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/evaluator/predicate/operator/comparison/InsensitiveBinaryComparison.java [hunk 16]
```diff
@@ -0,0 +1,31 @@
...
+    protected InsensitiveBinaryComparison(Source source, Expression left, Expression right) {
+        super(source, left, right);
+    }
+
+    @Override
+    public DataType dataType() {
+        return DataTypes.BOOLEAN;
+    }
+
+    public static boolean isWildcard(BytesRef value) {
```
Comment (astefan) [inline x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/evaluator/predicate/operator/comparison/InsensitiveBinaryComparison.java:27]: You probably already ruled this one out, but I'm mentioning it nevertheless.
In EQL the Optimizer is doing something that is in other few places where a rule is rewriting a specific operation with something that already exists in a different form already in the language. In SQL there is `Optimizer.ReplaceAggregatesWithLiterals` where `sum(1)` is replaced with an `iif`. For EQL it's `Optimizer.ReplaceWildcards` where the InsensitiveBinaryComparison logic is the same as the `Like` one.

So, the logic in ESQL for `=~` couldn't be, in part at least, replaced with what we already have for `LIKE`? We have a `RegexMatchEvaluator`, we have the `WildcardLike` that behaves exactly like `field_name =~ "*foo*"` so we could reuse this logic, no? With the Optimizer doing the replacement for `=~` in case the right hand side has wildcards, into a WildcardLike construct.

Comment (luigidellaquila) [reply x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/evaluator/predicate/operator/comparison/InsensitiveBinaryComparison.java:27]: TBH I think it should be the other way around, ie. 
- do not allow wildcards in `=~` (or better, do not treat them as wildcards, same as `==`)
- implement `LIKE~` (with wildcard support of course)
- replace `LIKE~` with `=~` if the pattern does not include wildcards, same as we already do for `LIKE` and `==`

Supporting wildcards in `=~` would make `LIKE~` completely redundant (there would be no difference between the two in terms of functionality). 
Moreover, if we support wildcards in `=~`, we will force users to escape wildcards in some cases (`?` in particular, that could be pretty common in text fields that contain natural language), and we know that escaping is painful.

Comment (luigidellaquila) [reply x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/evaluator/predicate/operator/comparison/InsensitiveBinaryComparison.java:27]: Your review is in line with the scope of this PR and with the original issue.
I just wanted to raise this concern here before moving forward.

I'd really like to have an opinion on the practical points (regardless of EQL, that is out of scope), in particular:
- what is the advantage of having two operators (`=~` and `LIKE~`) that do the exact same thing?
- we have `==` for exact match and `LIKE` for wildcard match, wouldn't it be more consistent to have `=~` for insensitive exact match and `LIKE~` for insensitive wildcard match?
- we have a rule that translates `LIKE` to `==` if there are no wildcards, why don't we do the same for `LIKE~` and `=~`?
- do we really want to force end users to escape `?` for exact match? 
  - for `==`: `{"query":"from idx | where x == \"how are you?\""}`
  - for `=~`: `{"query":"from idx | where x == \"how are you\\\\?\""}`

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Verifier.java (hunk 15 above)
[the remark this answers is not in the document]
Comment (luigidellaquila) [reply x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Verifier.java]: Ok, fixed. Since we are only accepting strings now, the checks are much easier. I just had to remove the Verifier logic

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/evaluator/predicate/operator/comparison/InsensitiveBinaryComparison.java (hunk 16 above)
Comment (costin) [reply x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/evaluator/predicate/operator/comparison/InsensitiveBinaryComparison.java:27]: @luigidellaquila please keep this PR focused on the items that are clear and move the rest of the items into follow-ups.
Namely let's trim this PR on just the first item in #103599:

> Introduce =~ operator for performing case insensitive equality between a string field and a string literal
The field can be both single and multi value. The literal has to be single value.

The multi-value aspect was descoped so let's get this PR to do:
> Introduce =~ operator for performing case insensitive equality between a string field and a string literal.
> The literal has to be single value.

Let's keep the wildcard, like~, etc.. for later.
Most of the code is already in - please clean it up, beef the testing, polish it up and then let's give it a final round of reviews.

For questions, let's talk offline.

Comment (luigidellaquila) [reply x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/evaluator/predicate/operator/comparison/InsensitiveBinaryComparison.java]: > Let's keep the wildcard, like~, etc.. for later.

I simplified the code removing wildcard logic. Now the scope is much more limited and IMHO it's self-contained enough, coherent with other operators (`==` logic in particular), and a good starting point for further evolution.

I added a few more tests, although the testable surface is much smaller with this simplification

**Comment by luigidellaquila** [comment]
@elasticmachine update branch

File: x-pack/plugin/esql/qa/testFixtures/src/main/resources/ignore-case.csv-spec
Comment (costin) [inline x-pack/plugin/esql/qa/testFixtures/src/main/resources/ignore-case.csv-spec:1]: It's not clear from the tests if the right hand side allows only literals (the original requirement), folded expressions or generic expressions (which you added).
I think it's the first variant but I don't see any tests validating this.
So please add more test to either validate that literals/folded expressions are required (and fields are not allowed) or vice-versa - queries that have fields on both sides and more over expressions:
`where concat(field, "constant") =~ concat(field, concat("con", "stant))` etc...

**Review by costin — COMMENTED** [review]
Thanks Luigi - this looks pretty complete.
I've added a comment regarding testing and whether field attributes are allowed or not both sides.

File: x-pack/plugin/esql/qa/testFixtures/src/main/resources/ignore-case.csv-spec
Comment (luigidellaquila) [reply x-pack/plugin/esql/qa/testFixtures/src/main/resources/ignore-case.csv-spec:1]: Thanks @costin 
Yes, the implementation supports any kind of expression, both on the left and on the right.
I added a few more tests for this.

File: x-pack/plugin/esql/qa/testFixtures/src/main/resources/ignore-case.csv-spec [hunk 17]
```diff
@@ -76,3 +76,36 @@ from employees | where first_name =~ "Georg?" | sort emp_no | keep emp_no, first
...
+expressionsLeft#[skip:-8.12.99, reason:case insensitive operators implemented in v 8.13]
+from employees | where concat(first_name, "_foo") =~ "TzvETAN_fOo" | keep emp_no, first_name;
+
+emp_no:integer  | first_name:keyword
+10007           | Tzvetan
+;
+
+
+expressionsLeftRight#[skip:-8.12.99, reason:case insensitive operators implemented in v 8.13]
+from employees | where substring(first_name, 1, 2) =~ substring(last_name, -2) | keep emp_no, first_name, last_name | sort emp_no;
```
Comment (costin) [inline x-pack/plugin/esql/qa/testFixtures/src/main/resources/ignore-case.csv-spec:106]: I've thought about this some more and we need to conclude how to address the missing requirements (multi-value and wildcard matching) from
https://github.com/elastic/elasticsearch/issues/103599 *before* enabling this functionality.
If `=~` is asymmetric like EQL `:` than comparing two fields can't happen. in ESQL we might do thing differently but until we reach a conclusion, I prefer we don't ship these features to begin with; it's better to enable them later than have them disabled.

In other words, please keep the comparison only between a literal/folded expression and a field but not between two fields.

**Review by costin — CHANGES_REQUESTED** [review]
Please add some tests between multi-values and single values.

And until we figure out #103599 fully when it comes to mv and wildcard behavior, please disable/prevent usage of this operator between fields in the Verifier.

File: x-pack/plugin/esql/qa/testFixtures/src/main/resources/ignore-case.csv-spec (hunk 17 above)
Comment (luigidellaquila) [reply x-pack/plugin/esql/qa/testFixtures/src/main/resources/ignore-case.csv-spec:106]: > in ESQL we might do thing differently but until we reach a conclusion, I prefer we don't ship these features to begin with

👍 I think it's a wise decision, I'll disable field vs field comparison at validation time (I'll allow only foldable expressions on the right) until we have a final decision on all the aspects.

**Comment by luigidellaquila** [comment]
@elasticmachine update branch

**Review by costin — APPROVED** [review]
Thanks for addressing the feedback - let's get this in!
